### DISCUSSION TITLE
Clarification on Neural Network Architecture with TrainWithClassifier Class

### DISCUSSION THREAD

**Discussion by learnfromgroundup** [KevinMusgrave/pytorch-metric-learning#655] [Q&A] [answered]
Hello Kevin and the community! I have a question regarding the neural network architecture when using TrainWithClassifier with triplet margin loss. I understand it takes a triplet of an anchor, positive, and negative image as input during training. However, the library abstracts away a lot of these details so elegantly that I wanted to confirm my understanding :)

If I were to draw a diagram of the neural network architecture, would it be correct to show the input as 3 images (anchor, positive, negative) that then pass through the CNN tracker, embedding layers, and classifier? Or does the library handle the triplet sampling under the hood in a way that the neural network itself just sees a batch of images as input?

I would greatly appreciate any clarification or confirmation on the architecture. Thank you!

**Comment by KevinMusgrave** [comment, marked as answer]
During both training and testing, the neural network just sees a batch of images. For each image, the neural network computes an embedding, independently of all the other images. You could have a batch size of 1, and the neural network wouldn't know the difference.

The triplet construction happens only to compute the loss value during training. Given the embeddings of a batch, the loss function is calculated by comparing the distances between anchor embedding, positive embedding, and negative embedding. But this is after the embeddings for the batch are computed. 

So the steps are:

1. Compute the embedding for each image in the batch
2. Compute the loss by finding all the triplets within the batch, and computing the distances between the embeddings that form those triplets
3. Update the neural network's parameters via backpropagation (calculating the gradient of the loss with respect to the final layer, the final layer with respect to the 2nd-last layer and so on).

The addition of the classifier layer just means that the triplet loss is computed after the 2nd-last layer instead of the final layer, and the classification loss is computed at the final layer.

**Comment by learnfromgroundup** [comment]
Hi Kevin, thanks for your detailed explanations. It is very clear.

As per my understanding, Triplet Loss and Siamese Neural Networks can also be used to learn a good distance function for the dataset. Do you have any intuitions about the similarities and differences between using Triplet Loss + Siamese Neural Networks and Triplet Loss + 'normal' CNN ?

**Reply by KevinMusgrave** [reply]
Triplet Loss + Siamese Neural Networks is the same as Triplet Loss + 'normal' CNN.

It's called Siamese Neural Network because the idea is you have two _identical copies_ of the network that take in different inputs, and then you compute the distance between the outputs.

But this is exactly the same as passing in all the inputs into a single network, and computing distances between the outputs of the single network. 

The "Siamese" naming is just to make the concept easier to understand.

**Reply by learnfromgroundup** [reply]
Thanks for explaining Kevin! I didn't hear anybody mentioning it. Thank you for your great insights :)
